### ISSUE TITLE
Cannot match return type `static` with return type `self` when subtype is `final`

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#5802]
In following example ( https://psalm.dev/r/0d757fabd6 ), psalm cannot match `self` in the `final class MyEvent` with the parent `static` return type definition:

```php
<?php

interface DomainEvent {
    /** @return static */
    public static function make(): self;
}

final class MyEvent implements DomainEvent
{
    public static function make(): self {
        return new self();
    }
}
```

```
Psalm output (using commit f62b83a): 

INFO: LessSpecificReturnStatement - 11:16 - The type 'MyEvent' is more general than the declared return type 'MyEvent&MyEvent&static' for MyEvent::make

INFO: MoreSpecificReturnType - 10:36 - The declared return type 'MyEvent&DomainEvent&static' for MyEvent::make is more specific than the inferred return type 'MyEvent'
```

When declaring `@return self` explicitly (at annotation level) no errors are raised ( https://psalm.dev/r/3f7b5b32fb )

```php
<?php

interface DomainEvent {
    /** @return static */
    public static function make(): self;
}

final class MyEvent implements DomainEvent
{
    /** @return self */
    public static function make(): self {
        return new self();
    }
}
```

Note that this also mostly applies to PHP 7.x codebases, since there's no way to use `static` explicitly there (declaring it works - https://psalm.dev/r/5ad7b7b05d ):

```php
<?php

interface DomainEvent {
    /** @return static */
    public static function make(): self;
}

final class MyEvent implements DomainEvent
{
    public static function make(): static {
        return new self();
    }
}
```

**Comment by Ocramius** [comment]
I think declaring `self` works well enough for my use-case: a more specific example would be needed to make this worth pursuing.

**Comment by calvinalkan** [comment]
I'm having a lot of issues with when using immutable objects:

This does not work.
https://psalm.dev/r/5c111b7d46

Changing :self to :View seems to work as expected:

https://psalm.dev/r/7abdb0d29e

**Comment by Ocramius** [comment]
It works when declaring `static` via native type declarations: https://psalm.dev/r/af3e198b4e

My suggestion is to go ahead and drop PHP 7 support, and just use the `static` native type.

**Comment by orklah** [comment]
Seems fixed :)

**Comment by Ocramius** [comment]
Awesome 💪
